backend/src/database/state-migrations/migrations.ts

```diff
@@ -136,71 +128,86 @@ export async function migrateInDatabaseTo(
+export function migrateStateExport(stateExport: StateExport): StateExport {
+    const { currentState, history } = applyMigrations(
+        stateExport.dataVersion,
+        stateExport.currentState,
+        stateExport.history
+            ? {
+                  initialState: stateExport.history.initialState,
+                  actions: stateExport.history.actionHistory,
+              }
+            : undefined
     );
-    migrationFunctions.state.forEach((migrateStateFunction) => {
-        stateExport.currentState = migrateStateFunction(
-            stateExport.currentState
-        ) as Mutable<ExerciseState>;
-        if (stateExport.history) {
-            stateExport.history.initialState = migrateStateFunction(
-                stateExport.history.initialState
-            ) as Mutable<ExerciseState>;
-        }
-    });
     if (stateExport.history) {
-        migrationFunctions.actions.forEach((fn) => {
-            fn(
-                stateExport.history!.initialState,
-                stateExport.history!.actionHistory
-            );
-        });
-        stateExport.history.actionHistory =
-            stateExport.history.actionHistory.filter(
-                (action) => action !== null
-            );
+        stateExport.history.initialState = history!
+            .initialState as Mutable<ExerciseState>;
+        stateExport.history.actionHistory = history!.actions.filter(
+            (action) => action !== null
+        ) as ExerciseAction[];
     }
+    stateExport.currentState = currentState as Mutable<ExerciseState>;
     return stateExport;
 }
 
-function getMigrationFunctions(
-    initialVersion: number,
-    targetVersion: number
-): { actions: MigrateActionsFunction[]; state: MigrateStateFunction[] } {
-    const necessaryMigrations = Object.entries(migrations)
-        .filter(
-            ([key]) =>
-                Number.parseInt(key) > initialVersion &&
-                Number.parseInt(key) <= targetVersion
-        )
-        .map(([, migration]) => migration);
+function applyMigrations(
+    currentStateVersion: number,
+    currentState: object,
+    history?: { initialState: object; actions: (object | null)[] }
+): {
+    currentStateVersion: number;
+    currentState: object;
+    history?: {
+        initialState: ExerciseState;
+        actions: (ExerciseAction | null)[];
+    };
+} {
+    const targetVersion = ExerciseState.currentStateVersion;
+    let newCurrentState = currentState;
+    for (let i = currentStateVersion + 1; i <= targetVersion; i++) {
+        const stateMigration = migrations[i]?.state;
+        if (typeof stateMigration === 'function') {
+            if (history)
+                history.initialState = stateMigration(history.initialState);
+            else newCurrentState = stateMigration(newCurrentState);
+        }
+        if (!history) continue;
+        const actionMigration = migrations[i]?.actions;
+        if (typeof actionMigration === 'function') {
```

Comment:
Checking for `typeof function` is weird. I propose to make `actions` and `state` optional (`?`) or a union with `undefined` in `Migration` instead of a union with `null`.
The migrations are never part of the state/get stringified, therefore I would not prefer `null` here (Sorry, if I previously proposed `null` for this).

Reply:
I'd keep the `null` because I want the person who writes a migration to explicitly say that some part of it is not required. I've changed the `?` to a `!` and then replaced this `typeof` check with `!== null` which also works.